// @azure/storage-file-share/dist-esm/src/generated/src/models/mappers.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
export var AccessPolicy = {
    serializedName: "AccessPolicy",
    type: {
        name: "Composite",
        className: "AccessPolicy",
        modelProperties: {
            startsOn: {
                xmlName: "Start",
                serializedName: "Start",
                type: {
                    name: "String"
                }
            },
            expiresOn: {
                xmlName: "Expiry",
                serializedName: "Expiry",
                type: {
                    name: "String"
                }
            },
            permissions: {
                xmlName: "Permission",
                serializedName: "Permission",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var CorsRule = {
    serializedName: "CorsRule",
    type: {
        name: "Composite",
        className: "CorsRule",
        modelProperties: {
            allowedOrigins: {
                xmlName: "AllowedOrigins",
                required: true,
                serializedName: "AllowedOrigins",
                type: {
                    name: "String"
                }
            },
            allowedMethods: {
                xmlName: "AllowedMethods",
                required: true,
                serializedName: "AllowedMethods",
                type: {
                    name: "String"
                }
            },
            allowedHeaders: {
                xmlName: "AllowedHeaders",
                required: true,
                serializedName: "AllowedHeaders",
                type: {
                    name: "String"
                }
            },
            exposedHeaders: {
                xmlName: "ExposedHeaders",
                required: true,
                serializedName: "ExposedHeaders",
                type: {
                    name: "String"
                }
            },
            maxAgeInSeconds: {
                xmlName: "MaxAgeInSeconds",
                required: true,
                serializedName: "MaxAgeInSeconds",
                constraints: {
                    InclusiveMinimum: 0
                },
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var DirectoryItem = {
    xmlName: "Directory",
    serializedName: "DirectoryItem",
    type: {
        name: "Composite",
        className: "DirectoryItem",
        modelProperties: {
            name: {
                xmlName: "Name",
                required: true,
                serializedName: "Name",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileProperty = {
    serializedName: "FileProperty",
    type: {
        name: "Composite",
        className: "FileProperty",
        modelProperties: {
            contentLength: {
                xmlName: "Content-Length",
                required: true,
                serializedName: "Content-Length",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var FileItem = {
    xmlName: "File",
    serializedName: "FileItem",
    type: {
        name: "Composite",
        className: "FileItem",
        modelProperties: {
            name: {
                xmlName: "Name",
                required: true,
                serializedName: "Name",
                type: {
                    name: "String"
                }
            },
            properties: {
                xmlName: "Properties",
                required: true,
                serializedName: "Properties",
                type: {
                    name: "Composite",
                    className: "FileProperty"
                }
            }
        }
    }
};
export var FilesAndDirectoriesListSegment = {
    xmlName: "Entries",
    serializedName: "FilesAndDirectoriesListSegment",
    type: {
        name: "Composite",
        className: "FilesAndDirectoriesListSegment",
        modelProperties: {
            directoryItems: {
                xmlName: "DirectoryItems",
                xmlElementName: "Directory",
                required: true,
                serializedName: "DirectoryItems",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "DirectoryItem"
                        }
                    }
                }
            },
            fileItems: {
                xmlName: "FileItems",
                xmlElementName: "File",
                required: true,
                serializedName: "FileItems",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "FileItem"
                        }
                    }
                }
            }
        }
    }
};
export var HandleItem = {
    xmlName: "Handle",
    serializedName: "HandleItem",
    type: {
        name: "Composite",
        className: "HandleItem",
        modelProperties: {
            handleId: {
                xmlName: "HandleId",
                required: true,
                serializedName: "HandleId",
                type: {
                    name: "String"
                }
            },
            path: {
                xmlName: "Path",
                required: true,
                serializedName: "Path",
                type: {
                    name: "String"
                }
            },
            fileId: {
                xmlName: "FileId",
                required: true,
                serializedName: "FileId",
                type: {
                    name: "String"
                }
            },
            parentId: {
                xmlName: "ParentId",
                serializedName: "ParentId",
                type: {
                    name: "String"
                }
            },
            sessionId: {
                xmlName: "SessionId",
                required: true,
                serializedName: "SessionId",
                type: {
                    name: "String"
                }
            },
            clientIp: {
                xmlName: "ClientIp",
                required: true,
                serializedName: "ClientIp",
                type: {
                    name: "String"
                }
            },
            openTime: {
                xmlName: "OpenTime",
                required: true,
                serializedName: "OpenTime",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            lastReconnectTime: {
                xmlName: "LastReconnectTime",
                serializedName: "LastReconnectTime",
                type: {
                    name: "DateTimeRfc1123"
                }
            }
        }
    }
};
export var ListFilesAndDirectoriesSegmentResponse = {
    xmlName: "EnumerationResults",
    serializedName: "ListFilesAndDirectoriesSegmentResponse",
    type: {
        name: "Composite",
        className: "ListFilesAndDirectoriesSegmentResponse",
        modelProperties: {
            serviceEndpoint: {
                xmlIsAttribute: true,
                xmlName: "ServiceEndpoint",
                required: true,
                serializedName: "ServiceEndpoint",
                type: {
                    name: "String"
                }
            },
            shareName: {
                xmlIsAttribute: true,
                xmlName: "ShareName",
                required: true,
                serializedName: "ShareName",
                type: {
                    name: "String"
                }
            },
            shareSnapshot: {
                xmlIsAttribute: true,
                xmlName: "ShareSnapshot",
                serializedName: "ShareSnapshot",
                type: {
                    name: "String"
                }
            },
            directoryPath: {
                xmlIsAttribute: true,
                xmlName: "DirectoryPath",
                required: true,
                serializedName: "DirectoryPath",
                type: {
                    name: "String"
                }
            },
            prefix: {
                xmlName: "Prefix",
                required: true,
                serializedName: "Prefix",
                type: {
                    name: "String"
                }
            },
            marker: {
                xmlName: "Marker",
                serializedName: "Marker",
                type: {
                    name: "String"
                }
            },
            maxResults: {
                xmlName: "MaxResults",
                serializedName: "MaxResults",
                type: {
                    name: "Number"
                }
            },
            segment: {
                xmlName: "Entries",
                required: true,
                serializedName: "Segment",
                type: {
                    name: "Composite",
                    className: "FilesAndDirectoriesListSegment"
                }
            },
            continuationToken: {
                xmlName: "NextMarker",
                required: true,
                serializedName: "NextMarker",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ListHandlesResponse = {
    xmlName: "EnumerationResults",
    serializedName: "ListHandlesResponse",
    type: {
        name: "Composite",
        className: "ListHandlesResponse",
        modelProperties: {
            handleList: {
                xmlIsWrapped: true,
                xmlName: "Entries",
                xmlElementName: "Handle",
                serializedName: "HandleList",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "HandleItem"
                        }
                    }
                }
            },
            continuationToken: {
                xmlName: "NextMarker",
                required: true,
                serializedName: "NextMarker",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareProperties = {
    serializedName: "ShareProperties",
    type: {
        name: "Composite",
        className: "ShareProperties",
        modelProperties: {
            lastModified: {
                xmlName: "Last-Modified",
                required: true,
                serializedName: "Last-Modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            etag: {
                xmlName: "Etag",
                required: true,
                serializedName: "Etag",
                type: {
                    name: "String"
                }
            },
            quota: {
                xmlName: "Quota",
                required: true,
                serializedName: "Quota",
                type: {
                    name: "Number"
                }
            },
            provisionedIops: {
                xmlName: "ProvisionedIops",
                serializedName: "ProvisionedIops",
                type: {
                    name: "Number"
                }
            },
            provisionedIngressMBps: {
                xmlName: "ProvisionedIngressMBps",
                serializedName: "ProvisionedIngressMBps",
                type: {
                    name: "Number"
                }
            },
            provisionedEgressMBps: {
                xmlName: "ProvisionedEgressMBps",
                serializedName: "ProvisionedEgressMBps",
                type: {
                    name: "Number"
                }
            },
            nextAllowedQuotaDowngradeTime: {
                xmlName: "NextAllowedQuotaDowngradeTime",
                serializedName: "NextAllowedQuotaDowngradeTime",
                type: {
                    name: "DateTimeRfc1123"
                }
            }
        }
    }
};
export var ShareItem = {
    xmlName: "Share",
    serializedName: "ShareItem",
    type: {
        name: "Composite",
        className: "ShareItem",
        modelProperties: {
            name: {
                xmlName: "Name",
                required: true,
                serializedName: "Name",
                type: {
                    name: "String"
                }
            },
            snapshot: {
                xmlName: "Snapshot",
                serializedName: "Snapshot",
                type: {
                    name: "String"
                }
            },
            properties: {
                xmlName: "Properties",
                required: true,
                serializedName: "Properties",
                type: {
                    name: "Composite",
                    className: "ShareProperties"
                }
            },
            metadata: {
                xmlName: "Metadata",
                serializedName: "Metadata",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                }
            }
        }
    }
};
export var ListSharesResponse = {
    xmlName: "EnumerationResults",
    serializedName: "ListSharesResponse",
    type: {
        name: "Composite",
        className: "ListSharesResponse",
        modelProperties: {
            serviceEndpoint: {
                xmlIsAttribute: true,
                xmlName: "ServiceEndpoint",
                required: true,
                serializedName: "ServiceEndpoint",
                type: {
                    name: "String"
                }
            },
            prefix: {
                xmlName: "Prefix",
                serializedName: "Prefix",
                type: {
                    name: "String"
                }
            },
            marker: {
                xmlName: "Marker",
                serializedName: "Marker",
                type: {
                    name: "String"
                }
            },
            maxResults: {
                xmlName: "MaxResults",
                serializedName: "MaxResults",
                type: {
                    name: "Number"
                }
            },
            shareItems: {
                xmlIsWrapped: true,
                xmlName: "Shares",
                xmlElementName: "Share",
                serializedName: "ShareItems",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "ShareItem"
                        }
                    }
                }
            },
            continuationToken: {
                xmlName: "NextMarker",
                required: true,
                serializedName: "NextMarker",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var RetentionPolicy = {
    serializedName: "RetentionPolicy",
    type: {
        name: "Composite",
        className: "RetentionPolicy",
        modelProperties: {
            enabled: {
                xmlName: "Enabled",
                required: true,
                serializedName: "Enabled",
                type: {
                    name: "Boolean"
                }
            },
            days: {
                xmlName: "Days",
                serializedName: "Days",
                constraints: {
                    InclusiveMaximum: 365,
                    InclusiveMinimum: 1
                },
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var Metrics = {
    serializedName: "Metrics",
    type: {
        name: "Composite",
        className: "Metrics",
        modelProperties: {
            version: {
                xmlName: "Version",
                required: true,
                serializedName: "Version",
                type: {
                    name: "String"
                }
            },
            enabled: {
                xmlName: "Enabled",
                required: true,
                serializedName: "Enabled",
                type: {
                    name: "Boolean"
                }
            },
            includeAPIs: {
                xmlName: "IncludeAPIs",
                serializedName: "IncludeAPIs",
                type: {
                    name: "Boolean"
                }
            },
            retentionPolicy: {
                xmlName: "RetentionPolicy",
                serializedName: "RetentionPolicy",
                type: {
                    name: "Composite",
                    className: "RetentionPolicy"
                }
            }
        }
    }
};
export var Range = {
    serializedName: "Range",
    type: {
        name: "Composite",
        className: "Range",
        modelProperties: {
            start: {
                xmlName: "Start",
                required: true,
                serializedName: "Start",
                type: {
                    name: "Number"
                }
            },
            end: {
                xmlName: "End",
                required: true,
                serializedName: "End",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var StorageError = {
    serializedName: "StorageError",
    type: {
        name: "Composite",
        className: "StorageError",
        modelProperties: {
            message: {
                xmlName: "Message",
                serializedName: "Message",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareStats = {
    serializedName: "ShareStats",
    type: {
        name: "Composite",
        className: "ShareStats",
        modelProperties: {
            shareUsageBytes: {
                xmlName: "ShareUsageBytes",
                required: true,
                serializedName: "ShareUsageBytes",
                type: {
                    name: "Number"
                }
            }
        }
    }
};
export var SignedIdentifier = {
    serializedName: "SignedIdentifier",
    type: {
        name: "Composite",
        className: "SignedIdentifier",
        modelProperties: {
            id: {
                xmlName: "Id",
                required: true,
                serializedName: "Id",
                type: {
                    name: "String"
                }
            },
            accessPolicy: {
                xmlName: "AccessPolicy",
                serializedName: "AccessPolicy",
                type: {
                    name: "Composite",
                    className: "AccessPolicy"
                }
            }
        }
    }
};
export var SharePermission = {
    serializedName: "SharePermission",
    type: {
        name: "Composite",
        className: "SharePermission",
        modelProperties: {
            permission: {
                xmlName: "permission",
                required: true,
                serializedName: "permission",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileServiceProperties = {
    xmlName: "StorageServiceProperties",
    serializedName: "FileServiceProperties",
    type: {
        name: "Composite",
        className: "FileServiceProperties",
        modelProperties: {
            hourMetrics: {
                xmlName: "HourMetrics",
                serializedName: "HourMetrics",
                type: {
                    name: "Composite",
                    className: "Metrics"
                }
            },
            minuteMetrics: {
                xmlName: "MinuteMetrics",
                serializedName: "MinuteMetrics",
                type: {
                    name: "Composite",
                    className: "Metrics"
                }
            },
            cors: {
                xmlIsWrapped: true,
                xmlName: "Cors",
                xmlElementName: "CorsRule",
                serializedName: "Cors",
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "CorsRule"
                        }
                    }
                }
            }
        }
    }
};
export var FileHttpHeaders = {
    xmlName: "fileHttpHeaders",
    type: {
        name: "Composite",
        className: "FileHttpHeaders",
        modelProperties: {
            fileContentType: {
                xmlName: "fileContentType",
                type: {
                    name: "String"
                }
            },
            fileContentEncoding: {
                xmlName: "fileContentEncoding",
                type: {
                    name: "String"
                }
            },
            fileContentLanguage: {
                xmlName: "fileContentLanguage",
                type: {
                    name: "String"
                }
            },
            fileCacheControl: {
                xmlName: "fileCacheControl",
                type: {
                    name: "String"
                }
            },
            fileContentMD5: {
                xmlName: "fileContentMD5",
                type: {
                    name: "ByteArray"
                }
            },
            fileContentDisposition: {
                xmlName: "fileContentDisposition",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var LeaseAccessConditions = {
    xmlName: "lease-access-conditions",
    type: {
        name: "Composite",
        className: "LeaseAccessConditions",
        modelProperties: {
            leaseId: {
                xmlName: "leaseId",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var SourceModifiedAccessConditions = {
    xmlName: "source-modified-access-conditions",
    type: {
        name: "Composite",
        className: "SourceModifiedAccessConditions",
        modelProperties: {
            sourceIfMatchCrc64: {
                xmlName: "sourceIfMatchCrc64",
                type: {
                    name: "ByteArray"
                }
            },
            sourceIfNoneMatchCrc64: {
                xmlName: "sourceIfNoneMatchCrc64",
                type: {
                    name: "ByteArray"
                }
            }
        }
    }
};
export var CopyFileSmbInfo = {
    xmlName: "copy-file-smb-info",
    type: {
        name: "Composite",
        className: "CopyFileSmbInfo",
        modelProperties: {
            filePermissionCopyMode: {
                xmlName: "filePermissionCopyMode",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "source",
                        "override"
                    ]
                }
            },
            ignoreReadOnly: {
                xmlName: "ignoreReadOnly",
                type: {
                    name: "Boolean"
                }
            },
            fileAttributes: {
                xmlName: "fileAttributes",
                type: {
                    name: "String"
                }
            },
            fileCreationTime: {
                xmlName: "fileCreationTime",
                type: {
                    name: "String"
                }
            },
            fileLastWriteTime: {
                xmlName: "fileLastWriteTime",
                type: {
                    name: "String"
                }
            },
            setArchiveAttribute: {
                xmlName: "setArchiveAttribute",
                type: {
                    name: "Boolean"
                }
            }
        }
    }
};
export var ServiceSetPropertiesHeaders = {
    serializedName: "service-setproperties-headers",
    type: {
        name: "Composite",
        className: "ServiceSetPropertiesHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ServiceGetPropertiesHeaders = {
    serializedName: "service-getproperties-headers",
    type: {
        name: "Composite",
        className: "ServiceGetPropertiesHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ServiceListSharesSegmentHeaders = {
    serializedName: "service-listsharessegment-headers",
    type: {
        name: "Composite",
        className: "ServiceListSharesSegmentHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareCreateHeaders = {
    serializedName: "share-create-headers",
    type: {
        name: "Composite",
        className: "ShareCreateHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareGetPropertiesHeaders = {
    serializedName: "share-getproperties-headers",
    type: {
        name: "Composite",
        className: "ShareGetPropertiesHeaders",
        modelProperties: {
            metadata: {
                serializedName: "x-ms-meta",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                },
                headerCollectionPrefix: "x-ms-meta-"
            },
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            quota: {
                serializedName: "x-ms-share-quota",
                type: {
                    name: "Number"
                }
            },
            provisionedIops: {
                serializedName: "x-ms-share-provisioned-iops",
                type: {
                    name: "Number"
                }
            },
            provisionedIngressMBps: {
                serializedName: "x-ms-share-provisioned-ingress-mbps",
                type: {
                    name: "Number"
                }
            },
            provisionedEgressMBps: {
                serializedName: "x-ms-share-provisioned-egress-mbps",
                type: {
                    name: "Number"
                }
            },
            nextAllowedQuotaDowngradeTime: {
                serializedName: "x-ms-share-next-allowed-quota-downgrade-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareDeleteHeaders = {
    serializedName: "share-delete-headers",
    type: {
        name: "Composite",
        className: "ShareDeleteHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareCreateSnapshotHeaders = {
    serializedName: "share-createsnapshot-headers",
    type: {
        name: "Composite",
        className: "ShareCreateSnapshotHeaders",
        modelProperties: {
            snapshot: {
                serializedName: "x-ms-snapshot",
                type: {
                    name: "String"
                }
            },
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareCreatePermissionHeaders = {
    serializedName: "share-createpermission-headers",
    type: {
        name: "Composite",
        className: "ShareCreatePermissionHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareGetPermissionHeaders = {
    serializedName: "share-getpermission-headers",
    type: {
        name: "Composite",
        className: "ShareGetPermissionHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareSetQuotaHeaders = {
    serializedName: "share-setquota-headers",
    type: {
        name: "Composite",
        className: "ShareSetQuotaHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareSetMetadataHeaders = {
    serializedName: "share-setmetadata-headers",
    type: {
        name: "Composite",
        className: "ShareSetMetadataHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareGetAccessPolicyHeaders = {
    serializedName: "share-getaccesspolicy-headers",
    type: {
        name: "Composite",
        className: "ShareGetAccessPolicyHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareSetAccessPolicyHeaders = {
    serializedName: "share-setaccesspolicy-headers",
    type: {
        name: "Composite",
        className: "ShareSetAccessPolicyHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var ShareGetStatisticsHeaders = {
    serializedName: "share-getstatistics-headers",
    type: {
        name: "Composite",
        className: "ShareGetStatisticsHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectoryCreateHeaders = {
    serializedName: "directory-create-headers",
    type: {
        name: "Composite",
        className: "DirectoryCreateHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            fileAttributes: {
                serializedName: "x-ms-file-attributes",
                type: {
                    name: "String"
                }
            },
            fileCreatedOn: {
                serializedName: "x-ms-file-creation-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileLastWriteOn: {
                serializedName: "x-ms-file-last-write-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileChangeOn: {
                serializedName: "x-ms-file-change-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileId: {
                serializedName: "x-ms-file-id",
                type: {
                    name: "String"
                }
            },
            fileParentId: {
                serializedName: "x-ms-file-parent-id",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectoryGetPropertiesHeaders = {
    serializedName: "directory-getproperties-headers",
    type: {
        name: "Composite",
        className: "DirectoryGetPropertiesHeaders",
        modelProperties: {
            metadata: {
                serializedName: "x-ms-meta",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                },
                headerCollectionPrefix: "x-ms-meta-"
            },
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            fileAttributes: {
                serializedName: "x-ms-file-attributes",
                type: {
                    name: "String"
                }
            },
            fileCreatedOn: {
                serializedName: "x-ms-file-creation-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileLastWriteOn: {
                serializedName: "x-ms-file-last-write-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileChangeOn: {
                serializedName: "x-ms-file-change-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            fileId: {
                serializedName: "x-ms-file-id",
                type: {
                    name: "String"
                }
            },
            fileParentId: {
                serializedName: "x-ms-file-parent-id",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectoryDeleteHeaders = {
    serializedName: "directory-delete-headers",
    type: {
        name: "Composite",
        className: "DirectoryDeleteHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectorySetPropertiesHeaders = {
    serializedName: "directory-setproperties-headers",
    type: {
        name: "Composite",
        className: "DirectorySetPropertiesHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            fileAttributes: {
                serializedName: "x-ms-file-attributes",
                type: {
                    name: "String"
                }
            },
            fileCreatedOn: {
                serializedName: "x-ms-file-creation-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileLastWriteOn: {
                serializedName: "x-ms-file-last-write-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileChangeOn: {
                serializedName: "x-ms-file-change-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileId: {
                serializedName: "x-ms-file-id",
                type: {
                    name: "String"
                }
            },
            fileParentId: {
                serializedName: "x-ms-file-parent-id",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectorySetMetadataHeaders = {
    serializedName: "directory-setmetadata-headers",
    type: {
        name: "Composite",
        className: "DirectorySetMetadataHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectoryListFilesAndDirectoriesSegmentHeaders = {
    serializedName: "directory-listfilesanddirectoriessegment-headers",
    type: {
        name: "Composite",
        className: "DirectoryListFilesAndDirectoriesSegmentHeaders",
        modelProperties: {
            contentType: {
                serializedName: "content-type",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectoryListHandlesHeaders = {
    serializedName: "directory-listhandles-headers",
    type: {
        name: "Composite",
        className: "DirectoryListHandlesHeaders",
        modelProperties: {
            contentType: {
                serializedName: "content-type",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var DirectoryForceCloseHandlesHeaders = {
    serializedName: "directory-forceclosehandles-headers",
    type: {
        name: "Composite",
        className: "DirectoryForceCloseHandlesHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            marker: {
                serializedName: "x-ms-marker",
                type: {
                    name: "String"
                }
            },
            numberOfHandlesClosed: {
                serializedName: "x-ms-number-of-handles-closed",
                type: {
                    name: "Number"
                }
            },
            numberOfHandlesFailedToClose: {
                serializedName: "x-ms-number-of-handles-failed",
                type: {
                    name: "Number"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileCreateHeaders = {
    serializedName: "file-create-headers",
    type: {
        name: "Composite",
        className: "FileCreateHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            fileAttributes: {
                serializedName: "x-ms-file-attributes",
                type: {
                    name: "String"
                }
            },
            fileCreatedOn: {
                serializedName: "x-ms-file-creation-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileLastWriteOn: {
                serializedName: "x-ms-file-last-write-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileChangeOn: {
                serializedName: "x-ms-file-change-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileId: {
                serializedName: "x-ms-file-id",
                type: {
                    name: "String"
                }
            },
            fileParentId: {
                serializedName: "x-ms-file-parent-id",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileDownloadHeaders = {
    serializedName: "file-download-headers",
    type: {
        name: "Composite",
        className: "FileDownloadHeaders",
        modelProperties: {
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            metadata: {
                serializedName: "x-ms-meta",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                },
                headerCollectionPrefix: "x-ms-meta-"
            },
            contentLength: {
                serializedName: "content-length",
                type: {
                    name: "Number"
                }
            },
            contentType: {
                serializedName: "content-type",
                type: {
                    name: "String"
                }
            },
            contentRange: {
                serializedName: "content-range",
                type: {
                    name: "String"
                }
            },
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            contentMD5: {
                serializedName: "content-md5",
                type: {
                    name: "ByteArray"
                }
            },
            contentEncoding: {
                serializedName: "content-encoding",
                type: {
                    name: "String"
                }
            },
            cacheControl: {
                serializedName: "cache-control",
                type: {
                    name: "String"
                }
            },
            contentDisposition: {
                serializedName: "content-disposition",
                type: {
                    name: "String"
                }
            },
            contentLanguage: {
                serializedName: "content-language",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            acceptRanges: {
                serializedName: "accept-ranges",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            copyCompletedOn: {
                serializedName: "x-ms-copy-completion-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            copyStatusDescription: {
                serializedName: "x-ms-copy-status-description",
                type: {
                    name: "String"
                }
            },
            copyId: {
                serializedName: "x-ms-copy-id",
                type: {
                    name: "String"
                }
            },
            copyProgress: {
                serializedName: "x-ms-copy-progress",
                type: {
                    name: "String"
                }
            },
            copySource: {
                serializedName: "x-ms-copy-source",
                type: {
                    name: "String"
                }
            },
            copyStatus: {
                serializedName: "x-ms-copy-status",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "pending",
                        "success",
                        "aborted",
                        "failed"
                    ]
                }
            },
            fileContentMD5: {
                serializedName: "x-ms-content-md5",
                type: {
                    name: "ByteArray"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            fileAttributes: {
                serializedName: "x-ms-file-attributes",
                type: {
                    name: "String"
                }
            },
            fileCreatedOn: {
                serializedName: "x-ms-file-creation-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileLastWriteOn: {
                serializedName: "x-ms-file-last-write-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileChangeOn: {
                serializedName: "x-ms-file-change-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            fileId: {
                serializedName: "x-ms-file-id",
                type: {
                    name: "String"
                }
            },
            fileParentId: {
                serializedName: "x-ms-file-parent-id",
                type: {
                    name: "String"
                }
            },
            leaseDuration: {
                serializedName: "x-ms-lease-duration",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "infinite",
                        "fixed"
                    ]
                }
            },
            leaseState: {
                serializedName: "x-ms-lease-state",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "available",
                        "leased",
                        "expired",
                        "breaking",
                        "broken"
                    ]
                }
            },
            leaseStatus: {
                serializedName: "x-ms-lease-status",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "locked",
                        "unlocked"
                    ]
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileGetPropertiesHeaders = {
    serializedName: "file-getproperties-headers",
    type: {
        name: "Composite",
        className: "FileGetPropertiesHeaders",
        modelProperties: {
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            metadata: {
                serializedName: "x-ms-meta",
                type: {
                    name: "Dictionary",
                    value: {
                        type: {
                            name: "String"
                        }
                    }
                },
                headerCollectionPrefix: "x-ms-meta-"
            },
            fileType: {
                serializedName: "x-ms-type",
                type: {
                    name: "String"
                }
            },
            contentLength: {
                serializedName: "content-length",
                type: {
                    name: "Number"
                }
            },
            contentType: {
                serializedName: "content-type",
                type: {
                    name: "String"
                }
            },
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            contentMD5: {
                serializedName: "content-md5",
                type: {
                    name: "ByteArray"
                }
            },
            contentEncoding: {
                serializedName: "content-encoding",
                type: {
                    name: "String"
                }
            },
            cacheControl: {
                serializedName: "cache-control",
                type: {
                    name: "String"
                }
            },
            contentDisposition: {
                serializedName: "content-disposition",
                type: {
                    name: "String"
                }
            },
            contentLanguage: {
                serializedName: "content-language",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            copyCompletedOn: {
                serializedName: "x-ms-copy-completion-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            copyStatusDescription: {
                serializedName: "x-ms-copy-status-description",
                type: {
                    name: "String"
                }
            },
            copyId: {
                serializedName: "x-ms-copy-id",
                type: {
                    name: "String"
                }
            },
            copyProgress: {
                serializedName: "x-ms-copy-progress",
                type: {
                    name: "String"
                }
            },
            copySource: {
                serializedName: "x-ms-copy-source",
                type: {
                    name: "String"
                }
            },
            copyStatus: {
                serializedName: "x-ms-copy-status",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "pending",
                        "success",
                        "aborted",
                        "failed"
                    ]
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            fileAttributes: {
                serializedName: "x-ms-file-attributes",
                type: {
                    name: "String"
                }
            },
            fileCreatedOn: {
                serializedName: "x-ms-file-creation-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileLastWriteOn: {
                serializedName: "x-ms-file-last-write-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileChangeOn: {
                serializedName: "x-ms-file-change-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            fileId: {
                serializedName: "x-ms-file-id",
                type: {
                    name: "String"
                }
            },
            fileParentId: {
                serializedName: "x-ms-file-parent-id",
                type: {
                    name: "String"
                }
            },
            leaseDuration: {
                serializedName: "x-ms-lease-duration",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "infinite",
                        "fixed"
                    ]
                }
            },
            leaseState: {
                serializedName: "x-ms-lease-state",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "available",
                        "leased",
                        "expired",
                        "breaking",
                        "broken"
                    ]
                }
            },
            leaseStatus: {
                serializedName: "x-ms-lease-status",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "locked",
                        "unlocked"
                    ]
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileDeleteHeaders = {
    serializedName: "file-delete-headers",
    type: {
        name: "Composite",
        className: "FileDeleteHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileSetHTTPHeadersHeaders = {
    serializedName: "file-sethttpheaders-headers",
    type: {
        name: "Composite",
        className: "FileSetHTTPHeadersHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            filePermissionKey: {
                serializedName: "x-ms-file-permission-key",
                type: {
                    name: "String"
                }
            },
            fileAttributes: {
                serializedName: "x-ms-file-attributes",
                type: {
                    name: "String"
                }
            },
            fileCreatedOn: {
                serializedName: "x-ms-file-creation-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileLastWriteOn: {
                serializedName: "x-ms-file-last-write-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileChangeOn: {
                serializedName: "x-ms-file-change-time",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            fileId: {
                serializedName: "x-ms-file-id",
                type: {
                    name: "String"
                }
            },
            fileParentId: {
                serializedName: "x-ms-file-parent-id",
                type: {
                    name: "String"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileSetMetadataHeaders = {
    serializedName: "file-setmetadata-headers",
    type: {
        name: "Composite",
        className: "FileSetMetadataHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileAcquireLeaseHeaders = {
    serializedName: "file-acquirelease-headers",
    type: {
        name: "Composite",
        className: "FileAcquireLeaseHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            leaseId: {
                serializedName: "x-ms-lease-id",
                type: {
                    name: "String"
                }
            },
            clientRequestId: {
                serializedName: "x-ms-client-request-id",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileReleaseLeaseHeaders = {
    serializedName: "file-releaselease-headers",
    type: {
        name: "Composite",
        className: "FileReleaseLeaseHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            clientRequestId: {
                serializedName: "x-ms-client-request-id",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileChangeLeaseHeaders = {
    serializedName: "file-changelease-headers",
    type: {
        name: "Composite",
        className: "FileChangeLeaseHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            leaseId: {
                serializedName: "x-ms-lease-id",
                type: {
                    name: "String"
                }
            },
            clientRequestId: {
                serializedName: "x-ms-client-request-id",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileBreakLeaseHeaders = {
    serializedName: "file-breaklease-headers",
    type: {
        name: "Composite",
        className: "FileBreakLeaseHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            leaseId: {
                serializedName: "x-ms-lease-id",
                type: {
                    name: "String"
                }
            },
            clientRequestId: {
                serializedName: "x-ms-client-request-id",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileUploadRangeHeaders = {
    serializedName: "file-uploadrange-headers",
    type: {
        name: "Composite",
        className: "FileUploadRangeHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            contentMD5: {
                serializedName: "content-md5",
                type: {
                    name: "ByteArray"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileUploadRangeFromURLHeaders = {
    serializedName: "file-uploadrangefromurl-headers",
    type: {
        name: "Composite",
        className: "FileUploadRangeFromURLHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            xMsContentCrc64: {
                serializedName: "x-ms-content-crc64",
                type: {
                    name: "ByteArray"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            isServerEncrypted: {
                serializedName: "x-ms-request-server-encrypted",
                type: {
                    name: "Boolean"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileGetRangeListHeaders = {
    serializedName: "file-getrangelist-headers",
    type: {
        name: "Composite",
        className: "FileGetRangeListHeaders",
        modelProperties: {
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            fileContentLength: {
                serializedName: "x-ms-content-length",
                type: {
                    name: "Number"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileStartCopyHeaders = {
    serializedName: "file-startcopy-headers",
    type: {
        name: "Composite",
        className: "FileStartCopyHeaders",
        modelProperties: {
            etag: {
                serializedName: "etag",
                type: {
                    name: "String"
                }
            },
            lastModified: {
                serializedName: "last-modified",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            copyId: {
                serializedName: "x-ms-copy-id",
                type: {
                    name: "String"
                }
            },
            copyStatus: {
                serializedName: "x-ms-copy-status",
                type: {
                    name: "Enum",
                    allowedValues: [
                        "pending",
                        "success",
                        "aborted",
                        "failed"
                    ]
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileAbortCopyHeaders = {
    serializedName: "file-abortcopy-headers",
    type: {
        name: "Composite",
        className: "FileAbortCopyHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileListHandlesHeaders = {
    serializedName: "file-listhandles-headers",
    type: {
        name: "Composite",
        className: "FileListHandlesHeaders",
        modelProperties: {
            contentType: {
                serializedName: "content-type",
                type: {
                    name: "String"
                }
            },
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
export var FileForceCloseHandlesHeaders = {
    serializedName: "file-forceclosehandles-headers",
    type: {
        name: "Composite",
        className: "FileForceCloseHandlesHeaders",
        modelProperties: {
            requestId: {
                serializedName: "x-ms-request-id",
                type: {
                    name: "String"
                }
            },
            version: {
                serializedName: "x-ms-version",
                type: {
                    name: "String"
                }
            },
            date: {
                serializedName: "date",
                type: {
                    name: "DateTimeRfc1123"
                }
            },
            marker: {
                serializedName: "x-ms-marker",
                type: {
                    name: "String"
                }
            },
            numberOfHandlesClosed: {
                serializedName: "x-ms-number-of-handles-closed",
                type: {
                    name: "Number"
                }
            },
            numberOfHandlesFailedToClose: {
                serializedName: "x-ms-number-of-handles-failed",
                type: {
                    name: "Number"
                }
            },
            errorCode: {
                serializedName: "x-ms-error-code",
                type: {
                    name: "String"
                }
            }
        }
    }
};
