// vscode-azureextensionui/out/src/AzExtOutputChannel.js
"use strict";
/*---------------------------------------------------------------------------------------------
*  Copyright (c) Microsoft Corporation. All rights reserved.
*  Licensed under the MIT License. See License.txt in the project root for license information.
*--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const vscode_1 = require("vscode");
// tslint:disable-next-line: export-name
function createAzExtOutputChannel(name, extensionPrefix) {
    return new AzExtOutputChannel(name, extensionPrefix);
}
exports.createAzExtOutputChannel = createAzExtOutputChannel;
class AzExtOutputChannel {
    constructor(name, extensionPrefix) {
        this.name = name;
        this.extensionPrefix = extensionPrefix;
        this._outputChannel = vscode_1.window.createOutputChannel(this.name);
    }
    append(value) {
        this._outputChannel.append(value);
    }
    appendLine(value) {
        this._outputChannel.appendLine(value);
    }
    appendLog(value, options) {
        const enableOutputTimestampsSetting = 'enableOutputTimestamps';
        const projectConfiguration = vscode_1.workspace.getConfiguration(this.extensionPrefix);
        const result = projectConfiguration.get(enableOutputTimestampsSetting);
        if (!result) {
            this.appendLine(value);
        }
        else {
            // tslint:disable: strict-boolean-expressions
            options = options || {};
            const date = options.date || new Date();
            this.appendLine(`${date.toLocaleTimeString()}${options.resourceName ? ' '.concat(options.resourceName) : ''}: ${value}`);
        }
    }
    clear() {
        this._outputChannel.clear();
    }
    // tslint:disable-next-line: no-any
    show(_column, preserveFocus) {
        this._outputChannel.show(preserveFocus);
    }
    hide() {
        this._outputChannel.hide();
    }
    dispose() {
        this._outputChannel.dispose();
    }
}
