// vscode-azureextensiondev/out/src/TestOutputChannel.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.md in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
class TestOutputChannel {
    constructor() {
        this.name = 'Extension Test Output';
    }
    append(value) {
        // Technically this is wrong (because of the new line), but good enough for now
        console.log(value);
    }
    appendLine(value) {
        console.log(value);
    }
    // tslint:disable-next-line: type-literal-delimiter
    appendLog(value, options) {
        // tslint:disable: strict-boolean-expressions no-parameter-reassignment
        options = options || {};
        const date = options.date || new Date();
        this.appendLine(`${date.toLocaleTimeString()}${options.resourceName ? ' '.concat(options.resourceName) : ''}: ${value}`);
    }
    clear() {
        // do nothing
    }
    show() {
        // do nothing
    }
    hide() {
        // do nothing
    }
    dispose() {
        // do nothing
    }
}
exports.TestOutputChannel = TestOutputChannel;
