// vscode-azureextensionui/out/src/wizard/StorageAccountCreateStep.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable-next-line:no-require-imports
const StorageManagementClient = require("azure-arm-storage");
const createAzureClient_1 = require("../createAzureClient");
const extensionVariables_1 = require("../extensionVariables");
const localize_1 = require("../localize");
const AzureWizardExecuteStep_1 = require("./AzureWizardExecuteStep");
class StorageAccountCreateStep extends AzureWizardExecuteStep_1.AzureWizardExecuteStep {
    constructor(defaults) {
        super();
        this.priority = 130;
        this._defaults = defaults;
    }
    execute(wizardContext, progress) {
        return __awaiter(this, void 0, void 0, function* () {
            // tslint:disable-next-line:no-non-null-assertion
            const newLocation = wizardContext.location.name;
            // tslint:disable-next-line:no-non-null-assertion
            const newName = wizardContext.newStorageAccountName;
            const newSkuName = `${this._defaults.performance}_${this._defaults.replication}`;
            const creatingStorageAccount = localize_1.localize('CreatingStorageAccount', 'Creating storage account "{0}" in location "{1}" with sku "{2}"...', newName, newLocation, newSkuName);
            extensionVariables_1.ext.outputChannel.appendLog(creatingStorageAccount);
            progress.report({ message: creatingStorageAccount });
            const storageClient = createAzureClient_1.createAzureClient(wizardContext, StorageManagementClient);
            wizardContext.storageAccount = yield storageClient.storageAccounts.create(
            // tslint:disable-next-line:no-non-null-assertion
            wizardContext.resourceGroup.name, newName, {
                sku: { name: newSkuName },
                kind: this._defaults.kind,
                location: newLocation,
                enableHttpsTrafficOnly: true
            });
            const createdStorageAccount = localize_1.localize('CreatedStorageAccount', 'Successfully created storage account "{0}".', newName);
            extensionVariables_1.ext.outputChannel.appendLog(createdStorageAccount);
        });
    }
    shouldExecute(wizardContext) {
        return !wizardContext.storageAccount;
    }
}
exports.StorageAccountCreateStep = StorageAccountCreateStep;
