// vscode-azureextensionui/out/src/treeDataProvider/GenericTreeItem.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const AzExtTreeItem_1 = require("./AzExtTreeItem");
class GenericTreeItem extends AzExtTreeItem_1.AzExtTreeItem {
    constructor(parent, options) {
        super(parent);
        this.label = options.label;
        this.contextValue = options.contextValue;
        this.id = options.id;
        this.commandId = options.commandId;
        this.iconPath = options.iconPath;
        this.description = options.description;
        this._includeInTreeItemPicker = !!options.includeInTreeItemPicker;
    }
    isAncestorOfImpl() {
        return this._includeInTreeItemPicker;
    }
}
exports.GenericTreeItem = GenericTreeItem;
